src/Adapter/Product/QueryHandler/GetProductCombinationsForEditingHandler.php

```diff
@@ -0,0 +1,149 @@
+<?php
+/**
+ * Copyright since 2007 PrestaShop SA and Contributors
+ * PrestaShop is an International Registered Trademark & Property of PrestaShop SA
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.md.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+declare(strict_types=1);
+
+namespace PrestaShop\PrestaShop\Adapter\Product\QueryHandler;
+
+use PrestaShop\PrestaShop\Adapter\Product\AbstractProductHandler;
+use PrestaShop\PrestaShop\Adapter\Product\Repository\CombinationRepository;
+use PrestaShop\PrestaShop\Adapter\Product\Repository\ProductRepository;
+use PrestaShop\PrestaShop\Core\Domain\Product\Combination\Query\GetProductCombinationsForEditing;
+use PrestaShop\PrestaShop\Core\Domain\Product\Combination\QueryHandler\GetProductCombinationsForEditingHandlerInterface;
+use PrestaShop\PrestaShop\Core\Domain\Product\Combination\QueryResult\CombinationAttributeInformation;
+use PrestaShop\PrestaShop\Core\Domain\Product\Combination\QueryResult\CombinationForEditing;
+use PrestaShop\PrestaShop\Core\Domain\Product\Combination\QueryResult\CombinationListForEditing;
+
+/**
+ * Handles @see GetProductCombinationsForEditing using legacy object model
+ */
+final class GetProductCombinationsForEditingHandler extends AbstractProductHandler implements GetProductCombinationsForEditingHandlerInterface
+{
+    /**
+     * @var CombinationRepository
+     */
+    private $combinationRepository;
+
+    /**
+     * @var ProductRepository
+     */
+    private $productRepository;
+
+    /**
+     * @param CombinationRepository $combinationRepository
+     * @param ProductRepository $productRepository
+     */
+    public function __construct(
+        CombinationRepository $combinationRepository,
+        ProductRepository $productRepository
+    ) {
+        $this->combinationRepository = $combinationRepository;
+        $this->productRepository = $productRepository;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function handle(GetProductCombinationsForEditing $query): CombinationListForEditing
+    {
+        $productId = $query->getProductId();
```

Comment:
You think it's better to do the check in the handler rather than in the `CombinationRepository` in the end?

Reply:
I don't really now.
Im wondering if/when repository methods should even hold those assertions, or should it be left to some upper layer.
Im afraid we will end up checking if product (or any other entity exists) multiple times per request if in each repository method we add assertion.
E.g. apply same principle in another PR (product duplicator) we are doing ~10 inserts to database with same product, so if we have a repository methods for each insertion & each of them would assert the same product over and over again it doesn't look right. :thinking: 
@matks maybe you could enlighten?

Reply:
Haha 😅  this is a gray area, it's hard to define the border between Handler and Repository.

So if we come back to the comparison for web standard MVC
- Handler is like a Web Controller
- Repository is like the Model in a MVC web app

If tomorrow I send a HTTP `GET /app/products/17`request to get product data, I expect a HTTP 404 response if product 17 does not exist. How will I do that ?
Well I believe we will have
- at controller level a check "does product 17 exists ?" and if it does not we return HTTP 404
- at model level when data is being fetched from database we should either check if product exists to avoid MySQL error or we should simply check if SQL query works correctly

So my answer would be
- yes, we need to check if product exists in the Handler, and if it does not exist we can exit early
- depending on the usecase, sometimes it's good to perform the check AGAIN in the repository (because method could be called from another place) following defensive programming strategy ; but not always